RecoTracker/MkFit/plugins/MkFitGeometryESProducer.cc

```diff
@@ -10,44 +11,430 @@
+  // Set some layer parameters (repeatedly, would require hard-coding otherwise)
+  layer_info.set_subdet(detid.subdetId());
+  layer_info.set_is_pixel(detid.subdetId() <= 2);
+  layer_info.set_is_stereo(m_trackerTopo->isStereo(detid));
+}
+
+//==============================================================================
+
+// Ideally these functions would also:
+// 0. Setup LayerInfo data (which is now done in auto-generated code).
+//    Some data-members are a bit over specific, esp/ bools for CMS sub-detectors.
+// 1. Establish short module ids (now done in MkFitGeometry constructor).
+// 2. Store module normal and strip direction vectors
+// 3. ? Any other information ?
+// 4. Extract stereo coverage holes where they exist (TEC, all but last 3 double-layers).
+//
+// Plugin DumpMkFitGeometry.cc can then be used to export this for stand-alone.
+// Would also need to be picked up with tk-ntuple converter (to get module ids as
+// they will now be used as indices into module info vectors).
+//
+// An attempt at export cmsRun config is in python/dumpMkFitGeometry.py
+
+void MkFitGeometryESProducer::addPixBGeometry(mkfit::TrackerInfo &trk_info) {
+  for (auto &det : m_trackerGeom->detsPXB()) {
+    fillShapeAndPlacement(det, trk_info);
+  }
+}
+
+void MkFitGeometryESProducer::addPixEGeometry(mkfit::TrackerInfo &trk_info) {
+  for (auto &det : m_trackerGeom->detsPXF()) {
+    fillShapeAndPlacement(det, trk_info);
+  }
+}
+
+void MkFitGeometryESProducer::addTIBGeometry(mkfit::TrackerInfo &trk_info) {
+  for (auto &det : m_trackerGeom->detsTIB()) {
+    fillShapeAndPlacement(det, trk_info);
+  }
+}
+
+void MkFitGeometryESProducer::addTOBGeometry(mkfit::TrackerInfo &trk_info) {
+  for (auto &det : m_trackerGeom->detsTOB()) {
+    fillShapeAndPlacement(det, trk_info);
+  }
+}
+
+void MkFitGeometryESProducer::addTIDGeometry(mkfit::TrackerInfo &trk_info) {
+  for (auto &det : m_trackerGeom->detsTID()) {
+    fillShapeAndPlacement(det, trk_info);
+  }
+}
+
+void MkFitGeometryESProducer::addTECGeometry(mkfit::TrackerInfo &trk_info) {
+  // For TEC we also need to discover hole in radial extents.
+  layer_gap_map_t lgc_map;
+  for (auto &det : m_trackerGeom->detsTEC()) {
+    fillShapeAndPlacement(det, trk_info, &lgc_map);
+  }
+  // Now loop over the GapCollectors and see if there is a coverage gap.
+  GapCollector::Interval itvl;
+  for (auto &[layer, gcol] : lgc_map) {
+    gcol.sqrt_elements();
+    if (gcol.find_gap(itvl, 0.5)) {
+      edm::LogInfo("TEC layer with gap") << "layer: " << layer << ", gap: " << itvl.x << " -> " << itvl.y
+                                         << " width = " << itvl.y - itvl.x;
+      trk_info.layer_nc(layer).set_r_hole_range(itvl.x, itvl.y);
+    }
+  }
+}
+
+//------------------------------------------------------------------------------
+// clang-format off
+namespace {
+  const float phase1QBins[] = {
```

Comment:
are these copied from somewhere else? perhaps I miss the point, but isn't this still hardcoding?

Reply:
Well, that's all that's left, there were many many more before, you know :)

These determine how hits are binned in z (barrel) and r (endcap) for fast lookup. In principle it's determined from module extent in that direction so we should be able to extract this during the loop over the modules as well. But this is potentially a larger physics change and might require also retuning of hit selection window parameters.